### PR TITLE
Add {clipboard} and {primary} to replace :paste

### PR DESCRIPTION
**Pull request by blyxxyz**
`:paste` is deprecated and replaced by equivalents using `:open` and the new variables, and `:open` supports opening newline-separated lists of URLs.

### CODE REVIEW

File: qutebrowser/config/configdata.py [hunk 1]
```diff
@@ -1512,12 +1512,12 @@ def data(readonly=False):
...
-        ('paste -t', ['Pp']),
-        ('paste -ts', ['PP']),
-        ('paste -w', ['wp']),
-        ('paste -ws', ['wP']),
+        ('open {clipboard}', ['pp']),
+        ('open {primary}', ['pP']),
+        ('open -t {clipboard}', ['Pp']),
+        ('open -t {primary}', ['PP']),
+        ('open -w {clipboard}', ['wp']),
+        ('open -w {primary}', ['wP']),
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.py]: Can you please add migrations for the existing bindings to the bottom of this file, and corresponding tests to `test_migrations` in `tests/unit/config/test_config.py`?

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -790,15 +816,9 @@ def paste(self, sel=False, tab=False, bg=False, window=False):
...
-        if sel and utils.supports_selection():
-            target = "Primary selection"
-        else:
+        if not utils.supports_selection():
             sel = False
-            target = "Clipboard"
         text = utils.get_clipboard(selection=sel)
-        if not text.strip():
-            raise cmdexc.CommandError("{} is empty.".format(target))
-        log.misc.debug("{} contained: {!r}".format(target, text))
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Why did you remove this?

Comment (blyxxyz) [inline qutebrowser/browser/commands.py]: I moved it to `get_clipboard`.

File: qutebrowser/utils/utils.py [hunk 3]
```diff
@@ -810,6 +811,11 @@ def get_clipboard(selection=False):
         mode = QClipboard.Selection if selection else QClipboard.Clipboard
         data = QApplication.clipboard().text(mode=mode)
 
+    target = "Primary selection" if selection else "Clipboard"
+    if not data.strip():
+        raise cmdexc.CommandError("{} is empty.".format(target))
```
Comment (The-Compiler) [inline qutebrowser/utils/utils.py]: Raising a `CommandError` from an utility function seems like a wrong thing to do - I'd prefer if you added a `ClipboardEmptyError` or so, raised that here, and then handled and re-raised that in the command handlers. Right now, this is also unhandled in `miscwidgets.py` where `utils.get_clipboard()` is used too.

Comment (blyxxyz) [inline qutebrowser/utils/utils.py]: Would a `CommandError` be appropriate to raise when parsing `{clipboard}` and `{primary}`?

**Comment by The-Compiler** [comment]
@hcraT Since you originally implemented multi-line pasting, want to help with reviewing this?

@blyxxyz If you could rebase this on the latest master and resolve conflicts that'd be appreciated - alternatively I can also do it while merging though.

File: qutebrowser/browser/commands.py [hunk 4]
```diff
@@ -245,37 +247,60 @@ def openurl(self, url=None, implicit=False,
...
-                if count is None:
-                    # We want to open a URL in the current tab, but none exists
-                    # yet.
-                    self._tabbed_browser.tabopen(url)
-                else:
-                    # Explicit count with a tab that doesn't exist.
-                    return
+            urllist = [u for u in url.split('\n') if u.strip()]
+            if (len(urllist) > 1 and
+                    any(not urlutils.is_url(u) and
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: This calls `urlutils.is_url` on all entries, which can possibly trigger a lot of DNS requests - I think that was the reason @hcraT only checked the first line in the original implementation, right?

File: qutebrowser/browser/commands.py [hunk 5]
```diff
@@ -245,37 +247,60 @@ def openurl(self, url=None, implicit=False,
...
+                        urlutils.get_path_if_valid(u, check_exists=True)
+                        is None for u in urllist)):
+                urllist = [url]
+                force_search = True
+            urls = [x for x in [self._parse_url(u, force_search=force_search)
+                                for u in urllist] if x is not None]
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: I find this hard to read - what about:

```
            urls = [self._parse_url(u, force_search=force_search)
                    for u in urllist]
            urls = [u for u in urllist if u is not None]
```

File: qutebrowser/browser/commands.py [hunk 6]
```diff
@@ -245,37 +247,60 @@ def openurl(self, url=None, implicit=False,
...
+        try:
+            return objreg.get('quickmark-manager').get(url)
+        except urlmarks.Error:
+            try:
+                return urlutils.fuzzy_url(url, force_search=force_search)
+            except urlutils.InvalidUrlError as e:
+                # We don't use cmdexc.CommandError here as this can be
+                # called async from edit_url
+                message.error(self._win_id, str(e))
+                return
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Please do `return None` explicitly here (as the other returns are explicit too)

File: qutebrowser/browser/commands.py [hunk 7]
```diff
@@ -245,37 +247,60 @@ def openurl(self, url=None, implicit=False,
...
+            if tab or bg or window:
+                self._open(cur_url, tab, bg, window, not implicit)
             else:
-                curtab.openurl(url)
+                curtab = self._cntwidget(count)
+                if curtab is None:
+                    if count is None:
+                        # We want to open a URL in the current tab, but none
+                        # exists yet.
+                        self._tabbed_browser.tabopen(cur_url)
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Please re-add the explicit `else` branch with the comment here again, otherwise I think this is somewhat confusing to read.

File: qutebrowser/browser/commands.py [hunk 8]
```diff
@@ -245,37 +247,60 @@ def openurl(self, url=None, implicit=False,
...
+                else:
+                    curtab.openurl(cur_url)
+
+    def _parse_url(self, url, force_search=False):
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: `force_search` could be a keyword-only argument (i.e., `self, url, *, force_search=False)`

File: qutebrowser/browser/commands.py (hunk 5 above)
Comment (blyxxyz) [inline qutebrowser/browser/commands.py]: Do you mean `urls = [u for u in urls if u is not None]`?

Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Err, yes I do. Sorry!

File: qutebrowser/browser/commands.py (hunk 7 above)
Comment (blyxxyz) [inline qutebrowser/browser/commands.py]: I removed it to comply with pylint's `too-many-branches`. Do you see a way to make the function simpler or should I just suppress it?

Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Maybe the whole multiline input parsing should be moved to a `_parse_url_input()` method or so, which yields URLs? This would also make it possible to write the loop/listcomp above:

```python
            urls = [self._parse_url(u, force_search=force_search)
                    for u in urllist]
            urls = [u for u in urllist if u is not None]
```

as:

```python
    for url in urllist:
        parsed = self._parse_url(u, force_search=force_search)
        if parsed is not None:
            yield parsed
```

File: qutebrowser/browser/commands.py (hunk 4 above)
Comment (hcraT) [inline qutebrowser/browser/commands.py]: It was thought as an heuristic. If the first line is an url probably all the lines contain an url otherwise the test is considered as an unique search query.

File: qutebrowser/commands/runners.py [hunk 9]
```diff
@@ -57,11 +57,19 @@ def replace_variables(win_id, arglist):
...
+        primary = utils.get_clipboard(selection=True)
     for arg in arglist:
         if arg == '{url}':
             args.append(url)
         elif arg == '{url:pretty}':
             args.append(pretty_url)
+        elif arg == '{clipboard}':
+            args.append(clipboard)
+        elif arg == '{primary}':
+            args.append(primary)
```
Comment (hcraT) [inline qutebrowser/commands/runners.py]: This code starts to look a bit convoluted (too many options each defining its own variable). In my opinion it should be refactored by using a dictionary (or similar declarative approach).

Comment (blyxxyz) [inline qutebrowser/commands/runners.py]: What's a good way of doing that without getting the clipboard contents if they're not needed and without getting the clipboard contents multiple times if the variable is used multiple times?

Comment (hcraT) [inline qutebrowser/commands/runners.py]: What about something like this:

```python
already_expanded = {}
args_expanders = {
  '{pretty:url}' : lambda: utils.get_clipboard(),
...
}
for arg in arglist:
  if arg not in already_expanded:
    already_expanded[arg] = args_expanders[ags]()
  args.append(already_expanded[arg])
```

File: qutebrowser/utils/utils.py (hunk 3 above)
Comment (The-Compiler) [inline qutebrowser/utils/utils.py]: Sorry, somehow missed this question. Since `replace_variables` gets called inside `run` and e.g. `run_safely` calls `run` and catches `cmdexc.CommandError` it should be alright.

File: qutebrowser/commands/runners.py (hunk 9 above)
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: Seems like we thought about the same :laughing: 

I guess @hcraT's approachj with iterating over `arglist` once and generating the values on the fly is a bit more straightforward than mine.

Comment (hcraT) [inline qutebrowser/commands/runners.py]: funny thing that we came up with essentially the same code! :)

**Comment by blyxxyz** [comment]
https://travis-ci.org/The-Compiler/qutebrowser/jobs/150943479 looks like a problem with Travis. The check also failed because of a docstring problem, I'll push a fix for that soon.

**Comment by The-Compiler** [comment]
Yeah, Travis had [some issues](https://www.traviscistatus.com/incidents/2p40l49r3yxd) (*"We've determined that an image cleanup process malfunctioned and deleted the stable build environment images"*...). I did rerun the build now and it passed.

**Comment by The-Compiler** [comment]
This is ready again from your POV, right?

File: qutebrowser/commands/runners.py [hunk 10]
```diff
@@ -49,21 +49,29 @@ def _current_url(tabbed_browser):
...
+
+    try:
+        for arg in arglist:
+            for var, func in variables.items():
+                if var in arg:
+                    if var not in values:
+                        values[var] = func()
+                    arg = arg.replace(var, values[var])
             args.append(arg)
+    except utils.ClipboardEmptyError as e:
```
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: I guess you should handle `utils.SelectionUnsupportedError` (i.e. using `:open {selection}` on Windows) here too? Probably time to make both a subclass of a `utils.ClipboardError` or so though, as we need to handle both in `miscwidgets.py` too.

File: qutebrowser/commands/runners.py [hunk 11]
```diff
@@ -49,21 +49,29 @@ def _current_url(tabbed_browser):
...
-        elif '{url:pretty}' in arg:
-            args.append(arg.replace('{url:pretty}', pretty_url))
-        else:
+
+    try:
+        for arg in arglist:
+            for var, func in variables.items():
+                if var in arg:
+                    if var not in values:
+                        values[var] = func()
```
Comment (hcraT) [inline qutebrowser/commands/runners.py]: Why do you cycle over the dictionary? If you are concerned about the possibility of receiving a variable that is not defined a simple test like this would work:

```python
for arg in arglist:
  if arg in variables:
    if var not in values:
      values[var] = variables[var]()
  else:
    raise WhateverException('{} is not an allowed variable placeholder'.format(var))
```

File: qutebrowser/commands/runners.py [hunk 12]
```diff
@@ -49,21 +49,29 @@ def _current_url(tabbed_browser):
...
+                    arg = arg.replace(var, values[var])
```
Comment (hcraT) [inline qutebrowser/commands/runners.py]: Why don't write just:

```python
args.append(arg.replace(var, values[var]))
```
Otherwise you are changing the arg variable without an apparent reason.

Comment (blyxxyz) [inline qutebrowser/commands/runners.py]: Arguments like `example.org/?a={url}&b={clipboard}` work this way.

File: qutebrowser/commands/runners.py (hunk 11 above)
Comment (blyxxyz) [inline qutebrowser/commands/runners.py]: Variables can be substrings of arguments.

Comment (hcraT) [inline qutebrowser/commands/runners.py]: Ok now all makes more sense.

In this case a neat thing could be to use format strings with named parameters. The only problem would be that you cannot use the colon in the name of the string.

Comment (blyxxyz) [inline qutebrowser/commands/runners.py]: Is there a way to dynamically change which named parameters `format` replaces? If not, creating a variable would become more work than just adding it to the `variables` dictionary.

File: qutebrowser/commands/runners.py (hunk 12 above)
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: @blyxxyz Is there a test doing this kind of thing? If not, there should be :wink:

File: qutebrowser/commands/runners.py (hunk 11 above)
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: I prefer this over using `str.format` because it's just much more predictable and simple :wink:

Comment (hcraT) [inline qutebrowser/commands/runners.py]: Not sure if this is better (this is untested):

```python
 def replace_variables(win_id, arglist):
     """Utility function to replace variables like {url} in a list of args."""
     """Utility function to replace variables like {url} in a list of args."""
    class __Replacer(dict):
      variables = {
           '{url}': lambda: _current_url(tabbed_browser).toString(
               QUrl.FullyEncoded | QUrl.RemovePassword),
           '{url:pretty}': lambda: _current_url(tabbed_browser).toString(
               QUrl.RemovePassword),
           '{clipboard}': utils.get_clipboard,
           '{primary}': lambda: utils.get_clipboard(selection=True),
       }

     def __init(self):
       self.__cached = {}

     def __getitem(self, key):
       if key in __Replacer.variables:
         if key not in self.__cached:
           self.__cached[key] = __Replacer.variables[key]()
         return self.__cached[key]
       else:
         raise WhateverException('Argument {} is not a valid placeholder'.format(key))

    replacer = __Replacer()
    args     = []
    formatter = string.Formatter()
    tabbed_browser = objreg.get('tabbed-browser', scope='window',
                                window=win_id)
                                window=win_id)
    for arg in arglist:
      args.append(formatter.vformat(arg, [], replacer))
```

What do you think?

Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: Looks overkill to me, and by subclassing from a builtin (`dict`) you're in for a world of surprises :wink:

**Comment by The-Compiler** [comment]
I'll give it another quick manual test before merging, but other than that it looks fine for me now, waiting for Travis to be finished.

**Comment by The-Compiler** [comment]
And merged - I broke Travis in master and didn't want to wait another round :laughing: 

Thanks a lot for the PR and the patience @blyxxyz, and thanks for the help with reviewing @hcraT!

**Comment by lahwaacz** [comment]
Sorry for coming late to this, but I think there should be a way to escape the pseudovariables for replacement - especially if more will be added in the future. I admit that it does not matter for commands that do some validation of the input in arguments, but e.g. `:open` is frequently used to pass unrecognized input to a search engine. Here the input can be anything (you never know what the compiler or interpreter spits out in the error messages :wink:) and it doesn't make much sense to replace if the result as a whole is not going to be recognized.

I think there is a regression that `{url}` is expanded everywhere, even in `:open foo {url}`, while previously only `:open {url}` was expanded. Similar problem is with `{clipboard}` and `{primary}`, but I don't see a clear way to handle it.

**Comment by The-Compiler** [comment]
Hmm, you're raising a good point... I wonder how much of a problem it will really be though, as long as we have a way to escape it somehow.

**Comment by blyxxyz** [comment]
Another problem is variables that get expanded into different variables. If your clipboard contains `{url}` and one of your arguments contains `{clipboard}` then the `{url}` gets expanded if that variable is checked later than `{clipboard}`.

**Comment by lahwaacz** [comment]
As for escaping, why don't we introduce the _void option_? Basically, anything after `--` in the arglist would be interpreted literally as positional arguments, whithout any replacement. Various tools (e.g. `tox`) use this trick for various purposes, e.g. passing arguments to wrapped tools or disambiguating file names beginning with `--`.

**Comment by The-Compiler** [comment]
@lahwaacz That sounds like a good idea, but I'm not sure if we can get the location of `--` from [`argparse`](https://docs.python.org/3.5/library/argparse.html) which we use to parse arguments. On the other hand, it should be trivial to get the location ourselves after splitting.

**Comment by lahwaacz** [comment]
That's cool, I just noticed that each command has an automatic `--help` option and `--` actually works in argparse to [separate positional arguments](https://docs.python.org/3.5/library/argparse.html#arguments-containing) :wink:

But the replacements are applied on the whole arglist, even before argparse comes into play. Consequently, things like `--{clipboard}` are expanded and if `{clipboard}` starts with `-` after expansion, the result will most likely be an error. I think the correct behaviour would be to expand only positionals after they are parsed with argparse, but the position of `--` (or more precisely the positionals to be expanded) will have to be determined manually.

**Comment by blyxxyz** [comment]
Using `--` to block both options and variables would be a bad idea. `--` is particularly useful for writing correct commands that expand variables because the variables might contain something that looks like an option. If your clipboard contains `-a` then pressing `pp` will fail. I'm fixing that at the moment.

**Comment by lahwaacz** [comment]
@blyxxyz I think the correct way to handle this is by doing the expansion of variables _after_ parsing the options with argparse. Then it doesn't matter if we use `--` for something more than argparse intended.

**Comment by The-Compiler** [comment]
Hmm, messing with the arguments after argparse parsed them sounds tricky. Even more so when you want to know where they came from, i.e. where there was a `--`.

**Comment by blyxxyz** [comment]
The most unixy way would be to not expand variables within (single) quotes, but I don't know if that would be any easier.

**Comment by The-Compiler** [comment]
We probably *could* get that information out from `qutebrowser.misc.split` if someone is crazy enough to try and mess with `ShellLexer` without breaking things (there are many good unittests though).

**Comment by lahwaacz** [comment]
Using (single) quotes might be the most unixy way, but that does not solve the problem with `:open`. With escaping via `--`, you could bind `o` to `:set-cmd-text -s :open -- ` to have the input "automatically" escaped, but with quotes the equivalent involves moving the cursor between the quotes, which is unnecessarily complicated.

**Comment by The-Compiler** [comment]
`:open` also doesn't take quotes :wink:

**Comment by lahwaacz** [comment]
Hmm, to make `:set-cmd-text -s :open -- ` work, we'd need to escape the `--` :sweat: I thought that we could use `:set-cmd-text -s -- :open -- ` to interpret the second `--` as a positional, but it seems that argparse does not like that.

**Comment by The-Compiler** [comment]
@lahwaacz #1861 please? :wink: (I edited the above comment after posting it though)
